### ISSUE TITLE
processing texts longer than 512 tokens with token-classification pipeline 

### ISSUE THREAD

**Issue by oliverguhr** [huggingface/transformers#15177]
# 🚀 Feature request

Currently, the token-classification pipeline truncates input texts longer than 512 tokens. 
It would be great if the pipeline could process texts of any length.

## Motivation

This issue is a result of this feature request:
https://github.com/huggingface/huggingface_hub/issues/532

As suggested I am tagging @Narsil here. 

## Your contribution

For my [punctuation prediction model](https://huggingface.co/oliverguhr/fullstop-punctuation-multilang-large), I wrote an inference code for this task.
This code segments a longer text into 512 tokens chunks with some overlap and 
reconstructs the final output string. The code however is rather complex and tailored to 
my punctuation prediction use case. So I am not sure if this is useful here.

**Comment by Narsil** [comment]
Hi @oliverguhr ,

So this specific pipelines truncates by default, and changing that would mean a breaking change.

The problem with "chunking" on this pipeline is that models usually need context before and after a token to get proper context to get a useful classification.
That means, than in order to achieve this, the pipeline should be able chunk into multiple parts of text, with overlap (usually called stride, at least in vision, but we borrowed that name in pipelines too). Run the `token-classification` on all the parts, and then resolve conflicts to give a final output.

Two things to have in mind:

- Conflict resolution is the tricky part and I don't think we can find a win always solution. It means that it's likely that the final result won't be as good as if we could have sent the full text to the model.
Consider the Example:

```
[The best city in the world is New York, because of Central Park]
--------------------------------------
                                ----------------------------------------
```

Both parts "see" New York, but one said it's a location while the other says it's nothing with greater confidence for instance. How would you resolve this ? What if New and York get different predictions for both parts. There's already a relatively complex resolution for tokens in the single case, this would make it more complex (that's ok).

- This can never be a default, always an opt-in. The pipelines' goal is to abstract away everything dealing with Tensors and logits, and manipulate simpler objects like strings and dicts. However, models being limited in range **is** a core limitation of those models (some models have an extremely long range) that we don't want to hide away (or at least not by default) especially if the result is not going to be the best one in all situations. Currently at least, we want users to know when there's a limitation and explicitly activate long range fallbacks. The only exception to the rule is `question-answering` currently.

**Comment by jammie19** [comment]
I've been looking for a similar feature to this, and also written custom, non-pipeline code to handle it so far.
I can't afford to throw away text past the max length, so use the stride as you describe to include additional context. This would be really handy to have included in the pipeline, given that the functionality to wrap overflowing tokens is in the tokenizer.

I agree that conflict resolution is the difficult part here, and will add complexity. I'd see it as similar to the way entities are aggregated from tokens. The pipeline would allow strategies to be passed, where this may use strategies such as "greatest confidence", "prefer entities", "prefer first" etc. (all pretty naive strategies, I know). This would let the user decide how truncation is, or isn't applied, with a flag to discard anything past the model max length (or some overriding max length).

I've been trying to implement this in a TokenClassificationPipeline subclass recently. If this truncation/overflowing were to be included in the pipeline, where would it go? Would it be part of the `preprocess` function? Or does it also need to be included in the `collate_fn` of the base? The `overflow_to_sample_mapping` key currently breaks the existing `collate_fn` as its shape is different, so something needs to change there.

**Comment by Narsil** [comment]
>  If this truncation/overflowing were to be included in the pipeline, where would it go? Would it be part of the preprocess function?

You would need to switch from `Pipeline` to `ChunkPipeline` (changes assumption of 1input-1 model pass, to 1 input-n model pass, while preserving batching, collation etc..).
Codewise this change is not too hard (going for `return` to `yield`, you can check `ZeroShotClassificationPipeline` for inspiration.
That should take care of the `collate_fn` for you (it's in `pipelines.pt_utils`) and `uncollate_fn` if you will too (since batching doesn't have to align with the number of items your input is creating). 

So preprocess would handle the chunking/overflowing+stride mecanisms, leveraging the tokenizer, but the post-process would be in charge of bringing the pieces together, `AutomaticSpeechRecognition` is another `ChunkPipeline` which has more postprocessing logic to stitch outputs.

Overall without actually digging into this work, it's hard to give you 100% confident advice on how to structure this. Aim for little code change and consistent with what's being done in the 2 other pipelines I have pointed out and it should be a good starting point.

If you start working on this, don't hesitate to share your work early and ping me or others to receive feedback on the direction.

**Comment by jammie19** [comment]
Thanks for the advice and links.
I see, I think that's where I've been going wrong. Ultimately, I want to be able to:
- make one call to `tokenize`, passing the entire batch at once
- Then apply batching before passing inputs to the `forward` function
- Then postprocess all inputs at once, such that the `overflow_to_sample_mapping` values are all correct and batch-independent (unaligned, as you say)

From the sounds of things, that's where the `collate_fn` and `uncollate_fn` functions come in. Does that seem like a sensible workflow?

**Comment by jammie19** [comment]
Great, that sounds perfect then (I should have looked at that before, it's been... difficult trying to hack that into a standard `Pipeline` subclass).

If I come up with anything useful and reusable, I'll open a PR

**Comment by Narsil** [comment]
> Great, that sounds perfect then (I should have looked at that before, it's been... difficult trying to hack that into a standard Pipeline subclass).

Perfectly understandable `ChunkPipeline` are relatively new, and mostly introduced after a big refactor, only to enable batching back onto `ZeroShot` and `QuestionAnswering`. It's not super advertised, since implementation is definitely more involved.

**Comment by drAbreu** [comment]
Hi all, I have come to a similar issue. My setup is I get the text of an image caption with several panels on it and I must separate the panels. 

My approach so far is using NER to classify tokens as 'O' or 'START-PANEL'. 

We use a BERT-based model but a substantial part of the captions we have are >512 tokens. I would be glad to give a try to build a kind of LongTextTokenClassificationPipeline following your comments if this has not yet been tried.

After experimenting with our model, in most of our cases `stride` would not be needed since the model keeps a high performance, but I could try to add something like that if the experience of buildiing this pipeline would be successful and useful for anyone

**Comment by Narsil** [comment]
> We use a BERT-based model but a substantial part of the captions we have are >512 tokens. I would be glad to give a try to build a kind of LongTextTokenClassificationPipeline following your comments if this has not yet been tried.

That would be awesome !

If you want this to be merged in `transformers` there is one key aspect, is that the chunking MUST be opt-in (because it will affect performance on models different from yours so we cannot do it without the users activating the option).

If you want to be easier, you could implement it exactly as you want and share it, and we could check what/how we can merge back afterwards.

**Comment by drAbreu** [comment]
Thanks for the answer :) 

I am following an approach that is basically using the `TokenClassificationPipeline` and modify it to be used with `ChunkPipeline` instead of `Pipeline`. 

So far things seem to be working. I am using the `stride` option for generalization. Will deal with how to aggregate the data when arriving to the `postprocess` part, but in the `preprocess` it does not generate a big issue. I am currently in the `preprocess` method. 

My current approach is to first use `_get_sentence_chunks` to estimate the total amount of chunks a sentence will be split on.
The next step is to divide the model inputs into these chunks. Here I am struggling a bit on what would be the most efficient way.

My thought was to set up `return_tensors=None` in the call to the tokenizer. This would give me lists that could be then sliced into the different parts of the sentence. Then, I need to add [CLS] and [END] to each chunk.

Then I would do a function (still not implemented here) that takes care of converting each list into a vector based on `self.framework` to make the pipeline general. 

Does this sound like a proper approach to you? 


`    
def preprocess(self, sentence, offset_mapping=None, stride=0):
        truncation = False
        
        model_inputs = self.tokenizer(
            sentence,
            return_tensors=None,
            truncation=truncation,
            return_special_tokens_mask=True,
            return_offsets_mapping=self.tokenizer.is_fast,
        )
        
        sentence_chunks = self._get_sentence_chunks(model_inputs["input_ids"], stride)
        print(sentence_chunks)
                        
        if offset_mapping:
            model_inputs["offset_mapping"] = offset_mapping
                    
        model_inputs["sentence"] = sentence
        bos_token = model_inputs["input_ids"][0]
        eos_token = model_inputs["input_ids"][-1]
        print(model_inputs.keys())
        for idx in range(sentence_chunks):
            start, end = self._get_start_end_idx(idx, model_inputs["input_ids"], stride)
            model_chunk_inputs = {
                'input_ids': [bos_token] + model_inputs['input_ids'][start:end] + [eos_token],
                'token_type_ids': [bos_token] + model_inputs['token_type_ids'][start:end] + [eos_token],
                'attention_mask': [bos_token] + model_inputs['attention_mask'][start:end] + [eos_token],
                'special_tokens_mask': [bos_token] + model_inputs['special_tokens_mask'][start:end] + [eos_token],
                'offset_mapping': [bos_token] + model_inputs['offset_mapping'][start:end] + [eos_token],
                'sentence': sentence,
                'sentence_chunk': tokenizer.decode([bos_token] + model_inputs['input_ids'][start:end] + [eos_token])
            }
            yield model_chunk_inputs
`

Just for completion, below are two extra methods used in 'preprocess'

`    
def _get_sentence_chunks(self, model_inputs: torch.Tensor, stride: int) -> int:
        length_sentence = model_inputs.size()[1]
        max_ = model.config.max_length - 2
        total_chunks = ceil(1 + (length_sentence - max_) / (max_ - stride))
        return total_chunks
    
  def _get_start_end_idx(self, idx: int, model_inputs: torch.Tensor, stride: int) -> Tuple[int, int]:
      length = model.config.max_length - 2
      start = idx * length - max(idx - 1, 0) * stride
      end = (idx + 1) * length - idx * stride
      return start, end
`

**Comment by jammie19** [comment]
Sorry for the radio silence @Narsil - I actually implemented this back in March, but due to issues with my employer's open source policy, couldn't publish the changes. These issues have been remedied, and I've got a branch with this token classification pipeline and associated tests.

I've implemented the pipeline, but had to migrate the tests from pytest to unittest tests so these tests don't all pass in the HF suite yet. I can open a (draft) PR at this stage to give you chance to comment? I'll try to finish those other test fixes in the next few days and publish the actual PR

**Comment by Narsil** [comment]
Sure, please publish ! 
Better to start the discussion early.

I don't think we hardly rely on `pytest` but I could be wrong.

**Comment by drAbreu** [comment]
Ok, I had been working on this the last , month too and was now on the tf tests. Should I then stop the work?

**Comment by jammie19** [comment]
I've gotten around to updating the test cases on the PR, they're passing now so I've published it. Let me know if it needs changes or explanation

**Comment by maxjeblick** [comment]
Any updates on this topic? 
Having a mechanism to handle more than 512 tokens for token classification inference is certainly a useful feature to have.

**Comment by jammie19** [comment]
I've got a [PR open](https://github.com/huggingface/transformers/pull/19735) which should solve this, but hasn't been reviewed yet. I've also had some issues with the CI tasks that I need someone from the HF team to help with
